src/pudl/metadata/descriptions.py

```diff
@@ -93,6 +94,34 @@
 """More standard descriptive text to appear in the Summary (first line) of resource descriptions, for timeseries resources."""
 
 
+def half_year_offset(partition: str, offset: int) -> str:
+    """Offset a half_year partition by the specified number of half_years."""
+    yr, _, half = partition.partition("half")
+    offset = pd.Timestamp(
+        year=int(yr), month=(int(half) - 1) * 6 + 1, day=1
+    ) + pd.DateOffset(months=offset * 6)
+    return f"{offset.year}half{offset.month / 6 + 1:.0f}"
+
+
+partition_offsets: dict[str, Callable] = {
+    "years": lambda partition, offset: str(
+        pd.Period(pd.Timestamp(partition) + pd.DateOffset(years=offset), freq="Y")
+    ).lower(),
+    "year_quarters": lambda partition, offset: str(
+        pd.Period(pd.Timestamp(partition) + pd.DateOffset(months=offset * 3), freq="Q")
+    ).lower(),
+    "half_years": half_year_offset,
+    "year_months": lambda partition, offset: str(
+        pd.Period(pd.Timestamp(partition) + pd.DateOffset(months=offset), freq="M")
+    ).lower(),
+}
+"""Lookup table for computing offsets of temporal partitions.
+
+Maps each partition key to a function that, given a partition (2026, 2026q3,
+2026half1, 2026-07) and an integer offset, returns the partition ``offset``
+partition-units away from the input partition."""
```

Comment:
These will need to be kept in sync with the temporal partition keys used by DataSource.

I'm pretty sure the description machinery is the only stuff that needs to do partition arithmetic, but if these would be useful elsewhere in the codebase lmk and I can file a followup issue to pull this into utils instead.